Nurumov_r/task02

robnur opened the pull request:
Не банте за 32 строчку пожалуйста
Полагаю реализация этого проекта предполагалась другая но мне в голову пришла такая, согласен немного с костылями но зато с изюминкой

UniFav committed:
new file:   Nurumov_r/task02/task02.py

robnur commented:
возможно мне придется все оформить в пакет с src tests итд но я пока не понял как мне это сделать в vs code, так же на PEP8 я не проверял, только визуально

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+from os import listdir
```
всё из os обычно импортируют без раскрытия namespace

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+from os import listdir
+from os.path import isdir
+
+def tree(lvl:int, path:str):
```
не соответствует PEP8

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+    lvl = lvl - 1
+    arr = listdir(path)
+    for e in sorted(arr):
+        if isdir(f"{path}/{e}"):
```
Объединение путей необходимо делать через os.path.join или использовать Path

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+    elements = []
+    lvl = lvl - 1
+    arr = listdir(path)
+    for e in sorted(arr):
```
Название переменной неудачное и непонятно, что означает

xthebat replied:
arr, почему не array? Не используйте бессмысленные сокращения

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+
+    return elements 
+
+def print_result(arrays, lvl):
```
нет аннотации типов

xthebat replied:
Функция print делает, что-то ещё кроме как print.
Зона ответственности функций должна быть согласна их названию и не выходить за её рамки.

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+
+def tree(lvl:int, path:str):
+    elements = []
+    lvl = lvl - 1
```
аналогично ниже - сокращения в называниях переменных

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+            files += count[1]
+            lvl -= 1
+        else:
+            if "\033[34m" in e:
```
константы выносим в начало функции или файла с именем

robnur replied:
я не понимаю про какую константу вы говорите

shadowk1337 replied:
`"\033[34m"`

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+                dirs += 1
+            else:
+                files += 1
+            print(lvl * "   " + e)
```
склейка строк - необходимо избавиться от неё

robnur replied:
как тогда реализовать умножение строк чтобы отобразить уровень вложенности

shadowk1337 replied:
джоин

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+            dirs += count[0]
+            files += count[1]
+            lvl -= 1
+        else:
```
Кажется, что можно использовать elif (забыл про него сказать на стриме)

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+
+    return elements 
+
+def print_result(arrays, lvl):
```
Выглядит, так что структуризация кода - неудачная. Должно быть как минимум две функции - одна собирает директории/файлы по заданной корневой директории в некую коллекцию (какую выбор за Вами). Вторая обходит её и выводит в заданный поток вывода с минимальной накрученной поверх логикой.

robnur replied:
Я хотел показать знание рекурсий(

robnur replied:
я правильно понял что вам не понравилась реализация? В функции tree я прохожу по директориям и записываю в массив все найденные директории и файлы (помечая их разными цветами) учитывая заданный уровень вложенности. В print_reult я вывожу все элементы массива подсчитывая директории и файлы (различать их я могу благодаря цвету), склейка строк мне нужна потому что я не знаю как еще можно использовать умножение строк, которое необходимо для визуального различия между разными уровнями вложенности

shadowk1337 replied:
Надо переделать структуру кода: выделить компонент, который отвечает за вывод, и выделить компонент, который отвечает за обработку объектов в директории и составляет их иерархию

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+def main(lvl:int, path:str):
+    mass = tree(lvl, path)
+    count = print_result(mass, 0)
+    print("\033[0m{}".format(f"\n{count[0]} directories, {count[1]} files"))
```
Magic'и в коде + формат смешанный с интерполяцией строк... Используем только интерполяцию строк для форматирования. Либо в обоснованных случаях что-то другое

robnur replied:
в данном случае не придумал как иначе сбросить стили и сделать вывод белым

xthebat replied:
Вынести Magic-и в переменные

robnur replied:
вы про это \033[0m{} ?

shadowk1337 replied:
Да, про это

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+    return dirs, files
+
+def main(lvl:int, path:str):
+    mass = tree(lvl, path)
```
Для чего здесь масса? Что это такое? ... Вообще это максимально неудачный нейминг. Сокращение + транслит. Используем только полное наименование без транслита.

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+            print(lvl * "   " + e)
+
+    return dirs, files
+
```
Нарушение PEP8

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+from os.path import isdir
+
+def tree(lvl:int, path:str):
+    elements = []
```
нет аннотации - это список чего

xthebat commented on Nurumov_r/task02/task02.py:
```diff
@@ -0,0 +1,42 @@
+    print("\033[0m{}".format(f"\n{count[0]} directories, {count[1]} files"))
+
+if __name__ == "__main__":
+    main(3,"./")
```
Не используется sys.argv

xthebat reviewed (changes requested):
Кроме указанных недочетов:
- отсутствуют тесты
- структуризация кода - не соответствует указанной на стриме
- PR должен быть выполнен с того аккаунта, который указан в Excel таблице

UniFav committed:
new file:   Nurumov_r/task02/src/task02.py
	deleted:    Nurumov_r/task02/task02.py

UniFav committed:
modified:   Nurumov_r/task02/src/task02.py

robnur commented:
Поправил что понял, проверку на PEP8 сделал, названия поменял, от склеек избавился

xthebat commented on Nurumov_r/task02/src/task02.py:
```diff
@@ -0,0 +1,50 @@
+    for element in sorted(array):
+        if os.path.isdir(os.path.join(path, element)):
+            if lvl > 0:
+                elements.insert(0, collect_tree(
```
120 символов длина строки по PEP8

xthebat commented on Nurumov_r/task02/src/task02.py:
```diff
@@ -0,0 +1,50 @@
+    for element in sorted(array):
+        if os.path.isdir(os.path.join(path, element)):
+            if lvl > 0:
+                elements.insert(0, collect_tree(
```
120 символов на строку

robnur replied:
я не понимаю о чем речь

shadowk1337 replied:
Речь о том, что PEP8 подразумевает, что в строке может быть до 120 символов, необязятельно тут переносить на следующую строку

xthebat commented:
Это не исправлено
- отсутствуют тесты
- структуризация кода - не соответствует указанной на стриме
- PR должен быть выполнен с того аккаунта, который указан в Excel таблице

shadowk1337 commented on Nurumov_r/task02/src/task02.py:
```diff
@@ -0,0 +1,50 @@
+
+def collect_tree(lvl: int, path: str):
+    elements = []
+    lvl = lvl - 1
```
```python
lvl -= 1
```

robnur replied:
не заметил, поправлю

shadowk1337 commented on Nurumov_r/task02/src/task02.py:
```diff
@@ -0,0 +1,50 @@
+            if lvl > 0:
+                elements.insert(0, collect_tree(
+                    lvl, os.path.join(path, element)))
+            elements.insert(0, f"\033[34m{element}")
```
Константы надо выносить, про это писали

robnur replied:
константы вы имеете ввиду "новый" путь?

shadowk1337 replied:
`\033[34m`

shadowk1337 commented on Nurumov_r/task02/src/task02.py:
```diff
@@ -0,0 +1,50 @@
+    dirs, files = 0, 0
+    for element in arrays:
+        if isinstance(element, list):
+            lvl += 1
```
ы? Зачем тут плюсовать, а в 29 строке минусовать `lvl`, если можно `tree` вызывать сразу как
```python
tree(element, lvl + 1)
```

robnur replied:
не подумал

shadowk1337 commented on Nurumov_r/task02/src/task02.py:
```diff
@@ -0,0 +1,50 @@
+    return elements
+
+
+def tree(arrays: list, lvl: int):
```
Нет типа возвращаемого значения

shadowk1337 commented on Nurumov_r/task02/src/task02.py:
```diff
@@ -0,0 +1,50 @@
+            files += count[1]
+            lvl -= 1
+        else:
+            if "\033[34m" in element:
```
Полагаться, на наличие такого символа вряд ли стоит. Окрашивание строк может убраться, тогда как определять, что `element` директория?

robnur replied:
мне пушить в массив вместо элемента массив где [0] это название файла/директории, а в [1] тип (файл это или директория)?

robnur replied:
да и я не совсем понимаю как оно может убраться до вывода если оно хранится в массиве. как по мне проверка на файл или директорию по цвету рациональное решение хоть и не совсем понятное/очевидное, могу комментарии написать

shadowk1337 reviewed (changes requested):
Нужно доделать логику и вывод команды. На данный момент команда не соответствует алгоритму работы команды `tree`:
- не выводится название директории от которой идет поиск
- не выводятся символы, обозначающие визуальную иерархию объектов в директории

robnur commented:
про 120 символов, документ сам так отформатировался под PEP8, могу конечно руками поправить но полагаю что если в следующий раз я снова отформатирую то могу забыть и будет снова такой перенос

robnur commented:
> Это не исправлено
> 
> * отсутствуют тесты
> * структуризация кода - не соответствует указанной на стриме
> * PR должен быть выполнен с того аккаунта, который указан в Excel таблице

тесты еще сделаю
про структуризацию кода не совсем понял, где можно посмотреть примеры? стрим я не записывал
аккаунт в таблице поправил

shadowk1337 commented:
> стрим я не записывал

в дискорде есть ссылка на запись лекции.......

UniFav committed:
deleted:    Nurumov_r/task02/src/task02.py
	new file:   Nurumov_r/task02/src/task02/__init__.py
	new file:   Nurumov_r/task02/src/task02/tree.py
	new file:   Nurumov_r/task02/test/test_tree.py

robnur commented:
> > стрим я не записывал
> 
> в дискорде есть ссылка на запись лекции.......

Не заметил сразу, искал в новых сообщениях

UniFav committed:
modified:   Nurumov_r/task02/src/task02/tree.py

UniFav committed:
modified:   Nurumov_r/task02/src/task02/tree.py

robnur commented:
Короче переписал весь код

UniFav committed:
modified:   Nurumov_r/task02/src/task02/tree.py

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+import os
+import sys
+
+directories_count = 0
+files_count = 0
+
+# Prefix components:
+space = '    '
+branch = '│   '
+tee = '├── '
+last = '└── '
+
+# Colors for console output
+white = '\033[0m'
+blue = '\033[94m'
+green = '\033[92m'
+red = '\033[31m'
```
Названия констант заглавными символами

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+    if len(sys.argv) > 2:
+        my_path = sys.argv[2]
+        try:
+            lvl = int(sys.argv[1]) - 1
+            if lvl < 0:
```
сокращения в названиях переменных

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def tree(directory: str, max_depth: int = 0, depth: int = 0, indent: str = ''):
+    global directories_count, files_count
```
Глобальные переменные - плохо, зачем они тебе тут?

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+if __name__ == "__main__":
+    if len(sys.argv) > 2:
+        my_path = sys.argv[2]
+        try:
```
К чему этот try?

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+red = '\033[31m'
+
+
+def tree(directory: str, max_depth: int = 0, depth: int = 0, indent: str = ''):
```
Как правило, названия функций содержат глагол

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+    for i, item in enumerate(directories):
+        path = os.path.join(directory, item)
+        if i == len(directories) - 1 and not files:
+            print(f"{white}{indent}{last}{blue}{item}")
```
Надо разделить обход директории и вывод в консоль

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+
+    for i, item in enumerate(directories):
+        path = os.path.join(directory, item)
+        if i == len(directories) - 1 and not files:
+            print(f"{white}{indent}{last}{blue}{item}")
+            tree(path, max_depth, depth + 1, f"{indent}{space}")
+        else:
+            print(f"{white}{indent}{tee}{blue}{item}")
+            tree(path, max_depth, depth + 1, f"{indent}{branch}")
```
Тут явное дублирование кода, такого лучше избегать

shadowk1337 commented on Nurumov_r/task02/src/task02/tree.py:
```diff
@@ -0,0 +1,81 @@
+
+    for i, item in enumerate(files):
+        path = os.path.join(directory, item)
+        if i == len(files) - 1:
+            print(f"{white}{indent}{last}{green}{item}")
+        else:
+            print(f"{white}{indent}{tee}{green}{item}")
```
дублирование кода

UniFav committed:
modified:   Nurumov_r/task02/src/task02/tree.py

robnur commented:
Вроде все поправил, я так понимаю осталось тесты сделать

shadowk1337 commented:
> Вроде все поправил, я так понимаю осталось тесты сделать

Остались незарезолвленные комментарии

robnur commented:
> > Вроде все поправил, я так понимаю осталось тесты сделать
> 
> Остались незарезолвленные комментарии

я не нахожу какие именно

UniFav committed:
modified:   Nurumov_r/task02/src/task02/tree.py
	new file:   Nurumov_r/task02/test/referents/referent_1.txt
	new file:   Nurumov_r/task02/test/referents/referent_2.txt
	new file:   Nurumov_r/task02/test/referents/referent_3.txt
	modified:   Nurumov_r/task02/test/test_tree.py

robnur commented:
Вроде выполнил все условия, в том числе и дополнительные

UniFav committed:
modified:   Nurumov_r/task02/src/task02/tree.py

xthebat reviewed (approved):
5/3
